clang/lib/CodeGen/CGCall.cpp

```diff
@@ -1672,10 +1672,11 @@ CodeGenTypes::GetFunctionType(const CGFunctionInfo &FI) {
 
   // Add type for sret argument.
   if (IRFunctionArgs.hasSRetArg()) {
-    QualType Ret = FI.getReturnType();
-    unsigned AddressSpace = CGM.getTypes().getTargetAddressSpace(Ret);
+    auto AddressSpace = CGM.getTarget().getIndirectArgAddressSpace();
```

Comment:
I would expect this to come from the ABIArgInfo/retAI, for the specific value and not a new side hook. Actually, is the address space already correct in retAI.getIndirectAddrSpace?

Reply:
Sadly no, that's not usable here as is, that's only for `byref` args (IndirectAliased). I do wonder if we should extend Indirect to also carry an AS, maybe that's the natural solution here.

Reply:
That's what I was thinking, yeah.  There should be plenty of space for that without inflating `ABIInfo`, right?

Reply:
I've taken a swipe at doing this; it's a bit noisier than I'd have hoped but it's mostly NFC except for AMDGPU, with other targets retaining current behaviour and having the option to adjust in the future. It does annoyingly add another relatively blind default use of AS0, but that was already there. Perhaps we can flip to defaulting to the AllocaAS with targets having the option to override the indirect AS when they classify a return type.